### PR TITLE
Fix some undefined behaviour

### LINKED ISSUE #3049: Undefined behaviour in snapshots
I found a funny bug again:

https://stackoverflow.com/questions/45429780/is-right-shifting-undefined-behavior-for-negative-number-in-cpp-and-in-java

I was profiling with MSVC under windows, and for some reason nameplates always called `LookupUuid`
which is quite inefficient (bcs its called for all clients every frame)

then i tried under linux.. couldn't reproduce.
then i analysed whats the difference between the two versions and found out that in snapshot.h
int Type() { return m_TypeAndID >> 16; }
does not return -1 under VC++ but 32767

### PR DESCRIPTION
**Pull request by def-**

### CODE REVIEW

**Comment by Jupeyy** [comment]
dunno if u want me to rebase on this pr, but
```
			int v = d[y * Pitch + x * 4];
			if(OrgWeight == 0)
				v = 0;
			else if(v <= OrgWeight)
				v = (int)(((v / (float)OrgWeight) * NewWeight));
			else if(InvOrgWeight == 0)
				v = NewWeight;
			else
				v = (int)(((v - OrgWeight) / (float)InvOrgWeight) * InvNewWeight + NewWeight);
			d[y * Pitch + x * 4] = v;
			d[y * Pitch + x * 4 + 1] = v;
			d[y * Pitch + x * 4 + 2] = v;```

should do the job

**Comment by def-** [comment]
> dunno if u want me to rebase on this pr, but this should do the job

Thanks, done.

File: src/base/math.h [hunk 1]
```diff
@@ -47,11 +47,11 @@ constexpr inline float fx2f(int v) { return v * (1.0f / (1 << 10)); }
 // int to fixed
 inline int i2fx(int v)
 {
-	return v << 10;
+	return v * 1024;
 }
 inline int fx2i(int v)
 {
-	return v >> 10;
+	return v / 1024;
```
Comment (heinrich5991) [inline src/base/math.h:54]: This does something different for negative integers, `-1 / 1024 == 0`, `-1 >> 10 == -1`.

File: src/engine/shared/snapshot.h [hunk 2]
```diff
@@ -13,7 +13,7 @@ class CSnapshotItem
 	int m_TypeAndID;
 
 	int *Data() { return (int *)(this + 1); }
-	int Type() { return m_TypeAndID >> 16; }
+	int Type() { return m_TypeAndID >= 0 ? m_TypeAndID >> 16 : -1; }
```
Comment (heinrich5991) [inline src/engine/shared/snapshot.h:16]: This changes behavior, did you verify that we don't rely on the old one?

**Review by heinrich5991 — CHANGES_REQUESTED** [review]
Rest looks good, thanks.

File: src/base/math.h (hunk 1 above)
Comment (def-) [reply src/base/math.h:54]: Done.

File: src/engine/shared/snapshot.h (hunk 2 above)
Comment (def-) [reply src/engine/shared/snapshot.h:16]: I'm not sure how to best fix this. Suggestions?

Comment (Jupeyy) [reply src/engine/shared/snapshot.h]: I can't see code where its relying on negative numbers except the -1, can you?

File: src/base/math.h (hunk 1 above)
Comment (heinrich5991) [reply src/base/math.h:59]: Okay, can we talk about the reason why you avoid right-shifting integers? Right-shifting negative integers is implementation-defined behavior until C++20 which is defined to do arithmetic right-shift in all major compilers.

https://github.com/ddnet/ddnet/issues/3049#issuecomment-706564322

The issue is not fixed by the branch, `>> 10` rounds towards -infinity, `/ 1024` rounds towards 0.

Comment (Jupeyy) [reply src/base/math.h:59]: but we are not at c++20 and msvc2017 failed

Comment (heinrich5991) [reply src/base/math.h:59]: Can you look at the assembly maybe?

Godbolt shows that right shifts are compiled to arithmetic right shifts by MSVC 2017:

https://godbolt.org/z/En8r5n

Comment (Jupeyy) [reply src/base/math.h:59]: yes, the assembly looks right, i'll try and see why the pointer is invalid then

File: src/engine/shared/snapshot.h (hunk 2 above)
Comment (Jupeyy) [reply src/engine/shared/snapshot.h]: image

Comment (Jupeyy) [reply src/engine/shared/snapshot.h]: Maybe bcs of the other snapshot problems, but my knowledge doesnt go deep here, its your turn now xd

Comment (heinrich5991) [reply src/engine/shared/snapshot.h]: > image

Not sure if that's what you're claiming, but
```
>>> "{:032b}".format(2147418112)
'01111111111111110000000000000000'
```
so 2147418112 is not the bit pattern of a negative number, and for positive numbers, right shift is strictly defined by the C standard.

Comment (Jupeyy) [reply src/engine/shared/snapshot.h]: yes, my claim is, that appearently the same snapshot has different values between linux and windows
So its probs caused by some of the other snapshot code

Comment (Jupeyy) [reply src/engine/shared/snapshot.h]: sry, maybe it was confusing that i replied to the other comment first, just wanted to switch here bcs this is the actual problem :D

**Comment by def-** [comment]
I removed the shift commits.

**Review by heinrich5991 — APPROVED** [review]
As said, looks good. Thanks for the fixes. :)

bors r+

**Pull request merged**